dbms/src/IO/Compression/CompressionCodecFOR.cpp

```diff
@@ -0,0 +1,254 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/BitpackingPrimitives.h>
+#include <Common/Exception.h>
+#include <IO/Compression/CompressionCodecFOR.h>
+#include <IO/Compression/CompressionInfo.h>
+#include <common/likely.h>
+#include <common/unaligned.h>
+
+#if defined(__AVX2__)
+#include <immintrin.h>
+#endif
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int CANNOT_COMPRESS;
+extern const int CANNOT_DECOMPRESS;
+} // namespace ErrorCodes
+
+CompressionCodecFOR::CompressionCodecFOR(UInt8 bytes_size_)
+    : bytes_size(bytes_size_)
+{}
+
+UInt8 CompressionCodecFOR::getMethodByte() const
+{
+    return static_cast<UInt8>(CompressionMethodByte::FOR);
+}
+
+UInt32 CompressionCodecFOR::getMaxCompressedDataSize(UInt32 uncompressed_size) const
+{
+    /**
+     *|bytes_of_original_type|frame_of_reference|width(bits)  |bitpacked data|
+     *|1 bytes               |bytes_size        |sizeof(UInt8)|required size |
+     */
+    const size_t count = uncompressed_size / bytes_size;
+    return 1 + bytes_size + sizeof(UInt8) + BitpackingPrimitives::getRequiredSize(count, bytes_size * 8);
+}
+
+template <std::integral T>
+UInt32 CompressionCodecFOR::compressData(const T * source, UInt32 count, char * dest)
+{
+    std::vector<T> values(count);
+    for (UInt32 i = 0; i < count; ++i)
+        values[i] = source[i];
+    T frame_of_reference = *std::min_element(values.cbegin(), values.cend());
```

Comment:
Do we need to handle `count == 0` in the previous code? Or it will make a crash here

Reply:
```
UInt32 CompressionCodecFOR::doCompressData(const char * source, UInt32 source_size, char * dest) const
{
    if unlikely (source_size % bytes_size != 0)
        throw Exception(ErrorCodes::CANNOT_DECOMPRESS, "source size {} is not aligned to {}", source_size, bytes_size);
```

if `count == 0`, will throw exception here.

Reply:
anyway, add a `RUNTIME_CHECK`